### ISSUE TITLE
first idea on pattern loops

### ISSUE THREAD

**Issue by leipert** [burodepeper/language-markdown#8]
As an alternative to our current header rules:
```coffeescript
key: 'heading'
containsLoops: true
loopData: {
  level: ['6','5','4','3','2','1']
}
patterns: [
  # Empty headings
  {
    match: '^(?: {0,3})((?:#{1,6})\\s*(?:#*)\\s*)$'
    name: 'empty.heading.md'
    captures:
      1:
        patterns: [
          {
            patternLoop: true,
            template: {
              match: '^(#{<%= level %>})\\s*(#*)\\s*$'
              name: 'heading-<%= level %>.md'
            }
            fixed: {
              captures: {
                1: { name: 'punctuation.md' }
                2: { name: 'punctuation.md' }
              }
            }
          }
        ]
  }
  {
    # NOTE
    # Does it make sense to first do a general check to determine if there is a
    # match for a heading? It seems more logical to test against one match, than
    # to go through all six of them every iteration.
    match: '^((?: {0,3})(?:#{1,6})(?: +)(?:.+))$'
    name: 'heading.md'
    captures:
      1:
        patterns: [
          {
            patternLoop: true,
            template: {
              begin: '^(?: {0,3})(#{<%= level %>})(?: +)'
              end: '(?: ){1}(#*)(?: *)$'
              name: 'heading-<%= level %>.md'
            }
            fixed: {
              captures:
                1: { name: 'punctuation.md' }
                2: patterns: [{ include: '#inline' }]
                3: { name: 'punctuation.md' }
            }
          }
        ]      
  }
]
```

**Comment by burodepeper** [comment]
I did not even know that was a possibility. Where do you find these things?

I was working on a different solution. In the `first-mate` package, there are two custom commands that can be used in the result of the match. One of them (`downcase`) is used in `language-todo`:

```cson
{
  'match': '(?<!\\w)@?(TODO|FIXME|CHANGED|XXX|IDEA|HACK|NOTE|REVIEW|NB)\\b'
  'name': 'storage.type.class.${1:/downcase}'
}
```

I've tried adding a `length` command, but because `first-mate` is a npm package, it works a little different from others, and I just can't get it to work in the way https://atom.io/docs/latest/behind-atom-developing-node-modules says it should.

My ideal solution would then look something like this:

```cson
key: 'heading'
patterns: [
  {
    match: '^(?: {0,3})((?:#{1,6})\\s*(?:#*)\\s*)$'
    name: 'empty.heading.md'
    captures:
      1:
        patterns: [
          {
            match: '^(#{1,6})\\s*(#*)\\s*$'
            name: 'heading-${1:/length}.md'
            captures: {
              1: { name: 'punctuation.md' }
              2: { name: 'punctuation.md' }
            }
        ]
  }
]
```

**Comment by leipert** [comment]
That is the beauty of it. It doesn't need to be possible.

We can make it possible, we have our own preprocessor(lib/main) which loads the rules, so there we can create custom stuff like looping.

We can do any addition to the rules (like disable, todo, loops with templates, debugging, etc).

**Comment by leipert** [comment]
So I do not keep finding these, I think of them ;)

**Comment by burodepeper** [comment]
Yeah, I already thought something along those lines. But would you mind trying to push that change into the first-mate package? I prefer to keep things 'basic' so others will understand things without any extra knowledge. But you can give it up after three sighs. I've sighed enough over that package.

**Comment by burodepeper** [comment]
@leipert Haven't heard from you in a long time, how are things?

I'm closing this issue/idea to keep things tidied up. I still think it is a good idea, but it might be better to implement these things as simple scripts (in a similar fashion that the fenced-code-blocks are generated) with a cson output, than as options that complicate the straightforwardedness of otherwise simple cson repositories. For a similar reason, I've also closed #10, although I think that actually concerns something that is missing from `cson-parse` or one of the other packages included.
